### ISSUE TITLE
Consensus rules to use dependency injection

### ISSUE THREAD

**Issue by dangershony** [stratisproject/StratisBitcoinFullNode#3487]
Currently the rules are in a weird state where they get registered on the protocol type initialization (POW/POW/POA), this is obviously wrong because rules are what define the networks ability to achieve consensus.

**Consensus Rules** are what represent the blockchain network itself, any change in rules will likely cause the node to not be able to validate transactions in a block and reject a block that is considered valid by the rest of the network because the block fails validation rules (hard fork, if the change makes a rule more restrictive its a soft fork and is okish) such changes will cause a regular node to not advance the chain and a mining node to fork away.

Most of the rules are stateless, however several rules need to access the store to function.
Just putting the rules in the Consensus class (as it is now partially, rules are in consensus but are registered at node bootstrap) will work but the rules will not be injectable (types the rule depend on will have to be injected in some initialize method or via some container, as is now by accessing the rules engine). ideally we want to be able to inject external component that a rule depends on (mainly the ability to access a resource like storage)

A proposal was made to register rule types in the consensus class, then inject the type to the DI service, and resolve the rules using the rules engine, this proposal will allow a rule to just use DI to access external components  

If rule types are initiated in the Network.Consensus class, we can then register them to the DI this way
```
    foreach (var ruleType in fullNodeBuilder.Network.Consensus.HeaderValidationRules)
        services.AddSingleton(typeof(IHeaderValidationConsensusRule), ruleType);

    foreach (var ruleType in fullNodeBuilder.Network.Consensus.IntegrityValidationRules)
        services.AddSingleton(typeof(IIntegrityValidationConsensusRule), ruleType);

    foreach (var ruleType in fullNodeBuilder.Network.Consensus.PartialValidationRules)
        services.AddSingleton(typeof(IPartialValidationConsensusRule), ruleType);

    foreach (var ruleType in fullNodeBuilder.Network.Consensus.FullValidationRules)
        services.AddSingleton(typeof(IFullValidationConsensusRule), ruleType);
```

The rules container declaration in IConsensus will be as following (I actually propose we move this to its own `ConsensusRules` class
```
        /// <summary>Group of rules that are used during block header validation specific to the given network.</summary>
        IEnumerable<Type> HeaderValidationRules { get; set; }

        /// <summary>Group of rules that are used during block integrity validation specific to the given network.</summary>
        IEnumerable<Type> IntegrityValidationRules { get; set; }

        /// <summary>Group of rules that are used during partial block validation specific to the given network.</summary>
        IEnumerable<Type> PartialValidationRules { get; set; }

        /// <summary>Group of rules that are used during full validation (connection of a new block) specific to the given network.</summary>
        IEnumerable<Type> FullValidationRules { get; set; }
```

To make sure rules order is maintained we can use IEnumerable feature with an anonymous method (alternatively we set an overridable method of the rules position on the rule itself but this is less preferred)
```
            IEnumerable<Type> HeaderValidationRules()
            {
                yield return typeof(HeaderTimeChecksRule);
            }

            consensus.HeaderValidationRules = HeaderValidationRules();
```

To resolve rules in the engine we should have a special type lets call it RulesContainer that will have collections of rules in its constructor and be used by the engine (on startup it will make sure the order of the rules is maintained)

**Comment by MithrilMan** [comment]
I'm fine with the proposal of IEnumerable method, so rule orders is clearly visible

so IConsensus should define 
`IEnumerable<Type> HeaderValidationRules { get; }` etc... but without the set (it shouldn't be modified from outside of IConsensusImplementation

then the IConsensusImplementation should populate that property with a method like

```
IEnumerable<Type> HeaderValidationRules()
{
    yield return typeof(HeaderTimeChecksRule);
    yield return typeof(AnotherRule);
    yield return typeof(AnotherDifferentRule);
}
```

Alternatively (probably better) we can get rid of the property and just have IConsensus defining
```
IEnumerable<Type> HeaderValidationRules()
IEnumerable<Type> IntegrityValidationRules()
IEnumerable<Type> PartialValidationRules()
IEnumerable<Type> FullValidationRules()
```

at the end this interface should be called just once during network initialization, so it's fine having a method that we can call directly from a NetworkBase class that every network has to inherit from

**Comment by fassadlr** [comment]
Fixed in #3497
